classes/metadata_fetcher.php

```diff
@@ -0,0 +1,112 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Utility class for fetching IDP metadata.
+ *
+ * @package    auth_saml2
+ * @author     Sam Chaffee
+ * @copyright  Copyright (c) 2017 Blackboard Inc. (http://www.blackboard.com)
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+namespace auth_saml2;
+
+defined('MOODLE_INTERNAL') || die();
+
+require_once(__DIR__ . '/../../../lib/filelib.php');
+
+/**
+ * Utility class for fetching IDP metadata.
+ *
+ * @package    auth_saml2
+ * @copyright  Copyright (c) 2017 Blackboard Inc. (http://www.blackboard.com)
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class metadata_fetcher {
+    /**
+     * @var array
+     */
+    private $curlinfo = [];
+
+    /**
+     * @var string
+     */
+    private $curlerror = '';
+
+    /**
+     * @var int
+     */
+    private $curlerrorno = 0;
+
+    /**
+     * @param $url
+     * @return bool
+     * @throws \coding_exception
+     */
+    public function fetch($url, $curl = null) {
+        if (!$curl instanceof \curl) {
+            $curl = new \curl();
+        }
+        $options = [
+            'CURLOPT_SSL_VERIFYPEER' => true,
```

Comment:
Made some changes related to this. The way I originally handled cURL errors should propagate the message the the failure was directly related to not being able to verify the peer or the host. I also used $CFG->forced_plugin_settings.

Unrelated to this, how would you like me to handle the unit test failures. They are all related to the fact that I'm using Prophecy to mock objects in the unit tests. That was added to PHPUnit in the version that was added to Moodle 3.0. I'd prefer not to change the use of Prophecy if possible.